ql/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorGroupByOperator.java

```diff
@@ -420,35 +460,56 @@ public void doProcessBatch(VectorizedRowBatch batch, boolean isFirstGroupingSet,
       //Flush if memory limits were reached
       // We keep flushing until the memory is under threshold
       int preFlushEntriesCount = numEntriesHashTable;
-      while (shouldFlush(batch)) {
-        flush(false);
 
-        if(gcCanary.get() == null) {
-          gcCanaryFlushes++;
-          gcCanary = new SoftReference<Object>(new Object());
-        }
+      if (!lruCache) {
+        while (shouldFlush(batch)) {
+          flush(false);
+
+          if(gcCanary.get() == null) {
+            gcCanaryFlushes++;
+            gcCanary = new SoftReference<Object>(new Object());
+          }
 
-        //Validate that some progress is being made
-        if (!(numEntriesHashTable < preFlushEntriesCount)) {
-          if (LOG.isDebugEnabled()) {
-            LOG.debug(String.format("Flush did not progress: %d entries before, %d entries after",
-                preFlushEntriesCount,
-                numEntriesHashTable));
+          //Validate that some progress is being made
+          if (!(numEntriesHashTable < preFlushEntriesCount)) {
+            if (LOG.isDebugEnabled()) {
+              LOG.debug(String.format("Flush did not progress: %d entries before, %d entries after",
+                  preFlushEntriesCount,
+                  numEntriesHashTable));
+            }
+            break;
           }
-          break;
+          preFlushEntriesCount = numEntriesHashTable;
         }
-        preFlushEntriesCount = numEntriesHashTable;
+      } else {
+        checkAndFlushLRU(batch);
       }
 
       if (sumBatchSize == 0 && 0 != batch.size) {
         // Sample the first batch processed for variable sizes.
         updateAvgVariableSize(batch);
+        currMemUsed = numEntriesHashTable * (fixedHashEntrySize + avgVariableSize);
```

Comment:
this is strange...there is a `currMemUsed` field an there is also a `currMemUsed` local variable in `shouldFlush` - they might cause things to me more interesting :)